### PR TITLE
Add rust implementation of consolidate_whitespace

### PR DESCRIPTION
**Pull request by zaowen**
You'll have to let me know if this is done the way we want.
I checked the current codebase and even though consolidate_whitespace has 2 arguments, one has a default arg which is never used. Because of this I just removed it from my implementation.

### CODE REVIEW

File: rust/libnewsboat-ffi/src/lib.rs [hunk 1]
```diff
@@ -35,6 +35,17 @@ pub extern "C" fn rs_replace_all(
...
 }
 
+#[no_mangle]
+pub extern "C" fn rs_consolidate_whitespace( input: *const c_char) -> *mut c_char {
+    let rs_input = unsafe { CStr::from_ptr(input) };
+    let rs_input = rs_input.to_string_lossy().into_owned();
+
+    let result = utils::consolidate_whitespace(rs_input);
+
+    let result = CString::new(result).unwrap();
```
Comment (Minoru) [inline rust/libnewsboat-ffi/src/lib.rs]: `unwrap` may panic, so please add a comment explaining why it won't in this case.

(I'd request a comment for the `unsafe` above, too, but: 1) don't feel entitled to do this since I didn't do it in replace_all; 2) I think I'll be moving that char*-to-String conversion into a function of its own, and I'll comment it there.)

File: rust/libnewsboat/src/utils.rs [hunk 2]
```diff
@@ -2,6 +2,30 @@ pub fn replace_all(input: String, from: &str, to: &str) -> String {
     input.replace(from, to)
 }
 
+pub fn consolidate_whitespace( input: String ) -> String {
+    let found = input.find( |c: char| !c.is_whitespace() );
+    let mut result = String::new();
+
+    if found.is_some() {
+        let (leading,rest) = input.split_at(found.unwrap());
```
Comment (Minoru) [inline rust/libnewsboat/src/utils.rs]: Please replace that `if` with `if let` so you don't have to call `unwrap`:

```rust
if let Some(found) = found {
    let (leading, rest) = input.split_at(found);
    ...
}
```

File: rust/libnewsboat/src/utils.rs [hunk 3]
```diff
@@ -2,6 +2,30 @@ pub fn replace_all(input: String, from: &str, to: &str) -> String {
...
+        let lastchar = input.chars().rev().next().unwrap();
```
Comment (Minoru) [inline rust/libnewsboat/src/utils.rs]: How can we be sure that this `unwrap` will never panic?

File: src/utils.cpp [hunk 4]
```diff
@@ -244,25 +244,9 @@ std::vector<std::string> utils::tokenize_spaced(const std::string& str,
 }
 
 std::string utils::consolidate_whitespace(const std::string& str,
-	std::string whitespace)
-{
-	std::string result;
-	std::string::size_type last_pos = str.find_first_not_of(whitespace);
-	std::string::size_type pos = str.find_first_of(whitespace, last_pos);
+	__attribute__((unused))std::string whitespace) {
```
Comment (Minoru) [inline src/utils.cpp]: Please make a separate commit that removes that argument altogether. If it's not used, it's a dead weight.

File: rust/libnewsboat/src/utils.rs (hunk 2 above)
Comment (zaowen) [reply rust/libnewsboat/src/utils.rs]: Thats a neat trick. Thank you.

File: rust/libnewsboat-ffi/src/lib.rs (hunk 1 above)
Comment (zaowen) [reply rust/libnewsboat-ffi/src/lib.rs]: Thanks.

**Comment by Minoru** [comment]
Thank you, @zaowen!
